From: Baolin Wang <baolin.wang@linux.alibaba.com>
Subject: mm: compaction: fix the possible deadlock when isolating hugetlb pages
Date: Mon, 13 Mar 2023 18:37:17 +0800

When trying to isolate a migratable pageblock, it can contain several
normal pages or several hugetlb pages (e.g.  CONT-PTE 64K hugetlb on
arm64) in a pageblock.  That means we may hold the lru lock of a normal
page to continue to isolate the next hugetlb page by
isolate_or_dissolve_huge_page() in the same migratable pageblock.

However in isolate_or_dissolve_huge_page(), it may allocate a new hugetlb
page and dissolve the old one via alloc_and_dissolve_hugetlb_folio() if
the hugetlb's refcount is zero.  That means we can still enter the direct
compaction path to allocate a new hugetlb page under the current lru lock,
which may cause a deadlock.

To avoid this possible deadlock, we should release the lru lock when
trying to isolate a hugetbl page.  Moreover it does not make sense to take
the lru lock to isolate a hugetlb, which is not in the lru list.

Link: https://lkml.kernel.org/r/a24a86fbae09711e61dc4424aa7aebff718e9995.1678703534.git.baolin.wang@linux.alibaba.com
Fixes: 369fa227c219 ("mm: make alloc_contig_range handle free hugetlb pages")
Signed-off-by: Baolin Wang <baolin.wang@linux.alibaba.com>
Reviewed-by: Mike Kravetz <mike.kravetz@oracle.com>
Reviewed-by: Vlastimil Babka <vbabka@suse.cz>
Cc: Mel Gorman <mgorman@techsingularity.net>
Cc: Oscar Salvador <osalvador@suse.de>
Cc: William Lam <william.lam@bytedance.com>
Signed-off-by: Andrew Morton <akpm@linux-foundation.org>
---


--- a/mm/compaction.c~mm-compaction-fix-the-possible-deadlock-when-isolating-hugetlb-pages
+++ a/mm/compaction.c
@@ -893,6 +893,11 @@ isolate_migratepages_block(struct compac
 		}
 
 		if (PageHuge(page) && cc->alloc_contig) {
+			if (locked) {
+				unlock_page_lruvec_irqrestore(locked, flags);
+				locked = NULL;
+			}
+
 			ret = isolate_or_dissolve_huge_page(page, &cc->migratepages);
 
 			/*
_
